apache/superset
# fix(explore): Prevent shared controls from checking feature flags outside React render

### SUMMARY

Feature flags are saved to a global window object and read using the `isFeatureEnabled` utility function, which returns `true` or `false` after checking the global window object.  As noted in #21201, this process is prone to error because feature flags are set at JS module execution time and sometimes also read at JS module execution time, creating a race condition that can be triggered by a change in webpack bundling.  This PR takes steps towards resolving one instance of this issue by moving the feature flag check to component render.  Previously, shared chart controls were exported as an object of control config objects, each of which contained the control's React component to render, params used by the control-rendering engine, and props that the control's React component would be rendered with.  For each shared control that had drag-and-drop functionality, a different config object would be exported at JS module execution time if `isFeatureEnabled` checks indicated that drag-and-drop functionality was enabled.  This PR moves the `isFeatureEnabled` check to inside a wrapper component, and uses a single config to render that wrapper component regardless of feature flag status.  While there was some overlap between config objects, the DnD and non-DnD versions of some control components required different props, so some refactoring was required to make it possible to merge the two config objects' props to form one unified config object.

### TESTING INSTRUCTIONS

- Confirm that every chart type's control panel works as expected when the `ENABLE_EXPLORE_DRAG_AND_DROP` feature flag is enabled and disabled.  When `ENABLE_EXPLORE_DRAG_AND_DROP` is enabled, also ensure that everything works when `ENABLE_DND_WITH_CLICK_UX` is enabled and disabled.

### ADDITIONAL INFORMATION


- [x] Required feature flags: `ENABLE_EXPLORE_DRAG_AND_DROP`, `ENABLE_DND_WITH_CLICK_UX`
